### DISCUSSION TITLE
How to create custom shapes/patterns of various sizes?

### DISCUSSION THREAD

**Discussion by jcg0474** [sn4k3/UVtools#772] [General]
Hello, I'm new to UVTools and have a sort of niche use for it. I'm patterning photoresist on a Silicon wafer and would like to create custom shapes/patterns of various sizes. I.e. an array of boxes that are 4x4 pixels, 8x8 pixels, 16x16 pixels, etc. I was able to do a single size with the pixel arithmetic tool, but I'd like to be able to do a lot of different sizes at once. I see that there is a "pattern" tool in the tools menu, however is is greyed out. I can't find anything about it in the documentation, is anyone familiar with it and why it would be greyed out? Thanks!

**Comment by sn4k3** [comment]
If anything is grey out means it's not compatible or not possible to use. In case of pattern is only grey if there is no space to pattern two times the current area as minimum.

To create custom shapes in exact place, run tools, workflows and all custom stuff at your will, you need to use UVtools scripting, see some [samples here](https://github.com/sn4k3/UVtools/tree/master/Scripts/UVtools.ScriptSample), in other words it's C# programing using UVtools Core, if you have any background with programming it may be easier.
You can also look at [ScriptTestPerLayerSettingsSample.cs](https://github.com/sn4k3/UVtools/blob/master/Scripts/UVtools.ScriptSample/ScriptTestPerLayerSettingsSample.cs) which creates a smile face under 5 layers

**Reply by jcg0474** [reply]
Thanks for the reply! By current area do you mean current selected area? Or the area of the entire layer? I'm still pretty new to the program so I don't quite understand everything. Is there a wiki explaining each feature, or is the only wiki available the one explaining the applications of the program? 
It's been a while since I've done any C# but I'll see what I can come up with.

**Reply by jcg0474** [reply]
Also, I don't really understand how to view the documentation. I see that is is a sandcastle builder help file, but I have never used anything like that and am struggling to view it in a meaningful way. Would you also be able to assist with this?

**Reply by sn4k3** [reply]
> By current area do you mean current selected area? Or the area of the entire layer?
the
The entire area of the model, the outer green rectangle that program show on the layers.

There is a wiki section here but only a few resources. There is no manual or complete wiki about the program. There is some youtube videos but more related to issues and test generation.


> Also, I don't really understand how to view the documentation. I see that is is a sandcastle builder help file, but I have never used anything like that and am struggling to view it in a meaningful way. Would you also be able to assist with this?

The documentation on the source to develop scripts you can see on the source itself or the nuget/fuget documentation however you need to explore the functions. You require to learn some of UVtools code to make the scripts.
I recommend you look into sample scritps, run them and see how they are doing the stuff, then try to do your functions. Draw itself its easy because you only need to call the function and give size and locations to draw shapes

**Reply by jcg0474** [reply]
Hi Tiago, thanks for replying. I ended up using Inkscape to generate the shapes that I need, its a tool I'm much more familiar with. However when I import the .tiff files I've made to create layers, I am unable to export them as a 3D printing file. I also cannot edit print parameters, it's greyed out. I'm assuming this has to do with the image not being in a format with the correct information to be recognized as a 3D file. I've also tried importing the images into a pre-existing 3D file, however I receive the error "Could not import and fit __ out of __ format(s) (Insufficient build space)." The resolution of the image matches the resolution of the build plate, so I'm not sure what's going on. Any help you could provide would be much appreciated, thanks!

**Reply by jcg0474** [reply]
I actually figured out the issue, it was pretty silly. For some reason when I was importing the image, it was being rotated. Thus the 1920x1080  image was trying to fit into a 1080x1920 space. I'm not sure why/how this was occurring, but it was an easy fix.

**Reply by sn4k3** [reply]
You need to create images matching the resolution that show on UVtools, status bar, bottom right.
